src/WebJobs.Script.WebHost/WebScriptHostManager.cs

```diff
@@ -349,6 +349,14 @@ protected override void OnHostStarted()
             _secretManager.PurgeOldFiles(Instance.ScriptConfig.RootScriptPath, Instance.TraceWriter);
         }
 
+        protected override void OnTimeout()
+        {
+            base.OnTimeout();
+            Stop();
```

Comment:
We should probably implement the IRegisteredObject pattern on WebScriptHostManager, and register our instance via HostingEnvironment.RegisterInstance. Then we'll get notified whenever shutdown is happening for any reason. There's also a web.config shutdownTimeout value that then takes effect, and we could set that there, rather than hardcoding 2000 inline here.

Kudu follows this pattern for some of its internal objects. It also has a [ShutdownDetector](https://github.com/projectkudu/kudu/blob/master/Kudu.Services/Infrastructure/ShutdownDetector.cs). @davidebbo  might know more on the history of that code and whether it would help us.